app/vlselect/traces/query/query.go

```diff
@@ -0,0 +1,404 @@
+func GetTraceList(ctx context.Context, cp *CommonParams, param *TraceQueryParam) ([]string, []*Row, error) {
+	currentTime := time.Now()
+
+	// query 1: * AND filter_confitions | last 1 by (_time) partition by (trace_id) | fields _time, trace_id | sort by (_time) desc
+	traceIDs, startTime, err := getTraceIDList(ctx, cp, param)
+	if err != nil {
+		return nil, nil, fmt.Errorf("get trace id error: %w", err)
+	}
+	if len(traceIDs) == 0 {
+		return nil, nil, nil
+	}
+
+	// query 2: trace_id:in(traceID, traceID, ...)
+	qStr := fmt.Sprintf(pb.TraceIDField+":in(%s)", strings.Join(traceIDs, ","))
+	q, err := logstorage.ParseQueryAtTimestamp(qStr, currentTime.UnixNano())
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot parse query [%s]: %s", qStr, err)
+	}
+
+	// adjust start time and end time with max duration window to make sure all spans are included.
+	q.AddTimeFilter(startTime.Add(-*traceMaxDurationWindow).UnixNano(), param.StartTimeMax.Add(*traceMaxDurationWindow).UnixNano())
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+
+	// search for trace spans and write to `rows []*Row`
+	var rowsLock sync.Mutex
+	var rows []*Row
+	var missingTimeColumn atomic.Bool
+	writeBlock := func(_ uint, db *logstorage.DataBlock) {
+		if missingTimeColumn.Load() {
+			return
+		}
+
+		columns := db.Columns
+		clonedColumnNames := make([]string, len(columns))
+		for i, c := range columns {
+			clonedColumnNames[i] = strings.Clone(c.Name)
+		}
+
+		timestamps, ok := db.GetTimestamps(nil)
+		if !ok {
+			missingTimeColumn.Store(true)
+			cancel()
+			return
+		}
+
+		for i, timestamp := range timestamps {
+			fields := make([]logstorage.Field, 0, len(columns))
+			for j := range columns {
+				// column could be empty if this span does not contain such field.
+				// only append non-empty columns.
+				if columns[j].Values[i] != "" {
+					fields = append(fields, logstorage.Field{Name: clonedColumnNames[j], Value: strings.Clone(columns[j].Values[i])})
+				}
+			}
+
+			rowsLock.Lock()
+			rows = append(rows, &Row{
+				Timestamp: timestamp,
+				Fields:    fields,
+			})
+			rowsLock.Unlock()
+		}
+	}
+
+	if err = vlstorage.RunQuery(ctxWithCancel, cp.TenantIDs, q, writeBlock); err != nil {
+		return nil, nil, err
+	}
+	if missingTimeColumn.Load() {
+		return nil, nil, fmt.Errorf("missing _time column in the result for the query [%s]", q)
+	}
+	return traceIDs, rows, nil
+}
+
+// getTraceIDList returns traceIDs according to the search params.
+// It also returns the earliest start time of these traces, to help reducing the time range for spans search.
+func getTraceIDList(ctx context.Context, cp *CommonParams, param *TraceQueryParam) ([]string, time.Time, error) {
+	currentTime := time.Now()
+	// query: * AND <filter> | last 1 by (_time) partition by (trace_id) | fields _time, trace_id | sort by (_time) desc
+	qStr := "*"
+	if param.ServiceName != "" {
+		qStr += fmt.Sprintf("AND _stream:{"+pb.ResourceAttrServiceName+"=\"%s\"} ", param.ServiceName)
```

Comment:
nit: instead of `\"%s\"` you can use `%q`.

Reply:
Thanks! updated